src/ripple/app/misc/impl/Manifest.cpp

```diff
@@ -27,44 +27,129 @@
 #include <ripple/json/json_reader.h>
 #include <ripple/protocol/PublicKey.h>
 #include <ripple/protocol/Sign.h>
+#include <boost/algorithm/clamp.hpp>
 #include <boost/regex.hpp>
 #include <numeric>
 #include <stdexcept>
 
 namespace ripple {
 
-boost::optional<Manifest>
-Manifest::make_Manifest (std::string s)
+boost::optional<Manifest> deserializeManifest(Slice s)
 {
+    if (s.empty())
+        return boost::none;
+
+    static SOTemplate const manifestFormat (
+        [](SOTemplate& t)
+        {
+            // A manifest must include:
+            // - master public key:
+            t.push_back (SOElement (sfPublicKey,       SOE_REQUIRED));
+
+            // - signature with that public key:
+            t.push_back (SOElement (sfMasterSignature, SOE_REQUIRED));
+
+            // - sequence number:
+            t.push_back (SOElement (sfSequence,        SOE_REQUIRED));
+
+            // It may, optionally, contain:
+            // - a version number which defaults to 0:
+            t.push_back (SOElement (sfVersion,          SOE_DEFAULT));
+
+            // - a domain name
+            t.push_back (SOElement (sfDomain,           SOE_OPTIONAL));
+
+            // - an ephemeral signing key that can be changed as necessary
+            t.push_back (SOElement (sfSigningPubKey,    SOE_OPTIONAL));
+
+            // - an signature using the ephemeral signing key, if it is present:
+            t.push_back (SOElement (sfSignature,        SOE_OPTIONAL));
+        });
+
     try
     {
-        STObject st (sfGeneric);
-        SerialIter sit (s.data (), s.size ());
-        st.set (sit);
+        SerialIter sit{ s };
+        STObject st{ sit, sfGeneric };
+
+        st.applyTemplate(manifestFormat);
+
+        // We only understand "version 0" manifests at this time:
+        if (st.isFieldPresent(sfVersion) && st.getFieldU16(sfVersion) != 0)
+            return boost::none;
+
         auto const pk = st.getFieldVL (sfPublicKey);
+
         if (! publicKeyType (makeSlice(pk)))
             return boost::none;
 
-        auto const opt_seq = get (st, sfSequence);
-        auto const opt_msig = get (st, sfMasterSignature);
-        if (!opt_seq || !opt_msig)
-            return boost::none;
+        Manifest m;
+        m.serialized.assign(reinterpret_cast<char const*>(s.data()), s.size());
```

Comment:
I agree that this would be ideal. I had wanted to include this in **1.2** and so sought to keep changes to a minimum, but the merge window closed on us, so we can be a bit more ambitious maybe.